File: js/src/cdOtimesCd/badziagPrivateSinglet.ts
/**
 * https://journals.aps.org/prresearch/abstract/10.1103/PhysRevResearch.3.023101
 *
 * P. Bądziąg, K. Horodecki, M. Horodecki, J. Jenkins, "Bound entanglement is
 * not sufficient for private key distillation"-style PPT singlets, presented
 * as ρ_F1 in Phys. Rev. Research 3, 023101 (2021).
 */

import { conj, multiply } from 'mathjs';
import type { Matrix } from 'mathjs';
import type { Scalar } from '../types.js';
import { fourier, ket, ketbra } from '../utils/index.js';
import { fail, sumMatrices } from '../utils/internal.js';

export interface BadziagPrivateSingletOptions {
  /**
   * The shield subsystem dimension d, at least 2. The state lives on
   * C² ⊗ C² ⊗ C^d ⊗ C^d in ABA'B' ordering, i.e. C^2d ⊗ C^2d across the
   * Alice|Bob cut once the systems are reordered to AA'BB'.
   */
  shieldDim: number;
}

/** The product basis state |i j k l> of C² ⊗ C² ⊗ C^d ⊗ C^d. */
function basis(d: number, i: number, j: number, k: number, l: number): Matrix {
  return ket([2, 2, d, d], [i, j, k, l]);
}

/**
 * The Bądziąg et al. private-singlet bound entangled state on C^2d ⊗ C^2d.
 *
 * A PPT singlet built from a shield pair A'B' of local dimension d and a
 * two-qubit pair AB, mixed with weights `p1 = √d / (1 + √d)` and `p2 = 1 - p1`.
 * The state is bound entangled for every `shieldDim >= 2`. The returned matrix
 * is in ABA'B' ordering, so testing the Alice|Bob cut requires permuting the
 * systems to AA'BB' first.
 *
 * @param options - see {@link BadziagPrivateSingletOptions}.
 * @returns the (4d²) x (4d²) density matrix of the private singlet.
 */
export function badziagPrivateSinglet({ shieldDim }: BadziagPrivateSingletOptions): Matrix {
  const d = shieldDim;
  if (!Number.isInteger(d) || d < 2) {
    fail(`the shield dimension must be an integer >= 2, got ${d}`);
  }

  const p1 = Math.sqrt(d) / (1 + Math.sqrt(d));
  const p2 = 1 - p1;
  const u = fourier(d);

  // The diagonal |00 ij> and |11 ij> populations of the p1 branch.
  const diagonal00And11: Matrix[] = [];
  for (let i = 0; i < d; i++) {
    for (let j = 0; j < d; j++) {
      diagonal00And11.push(ketbra(basis(d, 0, 0, i, j)));
      diagonal00And11.push(ketbra(basis(d, 1, 1, i, j)));
    }
  }

  // The |00 ij><11 ji| coherences of the p1 branch, weighted by the Fourier
  // matrix, plus their Hermitian conjugates.
  const coherence00To11: Matrix[] = [];
  for (let i = 0; i < d; i++) {
    for (let j = 0; j < d; j++) {
      const uij = u.get([i, j]) as Scalar;
      coherence00To11.push(
        multiply(uij, ketbra(basis(d, 0, 0, i, j), basis(d, 1, 1, j, i))) as Matrix,
      );
      coherence00To11.push(
        multiply(conj(uij), ketbra(basis(d, 1, 1, j, i), basis(d, 0, 0, i, j))) as Matrix,
      );
    }
  }

  // The diagonal |01 ii> and |10 ii> populations of the p2 branch.
  const diagonal01And10: Matrix[] = [];
  for (let i = 0; i < d; i++) {
    diagonal01And10.push(ketbra(basis(d, 0, 1, i, i)));
    diagonal01And10.push(ketbra(basis(d, 1, 0, i, i)));
  }

  // The |01 ii><10 jj| coherences of the p2 branch and their conjugates.
  const coherence01To10: Matrix[] = [];
  for (let i = 0; i < d; i++) {
    for (let j = 0; j < d; j++) {
      const uij = u.get([i, j]) as Scalar;
      coherence01To10.push(
        multiply(uij, ketbra(basis(d, 0, 1, i, i), basis(d, 1, 0, j, j))) as Matrix,
      );
      coherence01To10.push(
        multiply(conj(uij), ketbra(basis(d, 1, 0, j, j), basis(d, 0, 1, i, i))) as Matrix,
      );
    }
  }

  return sumMatrices([
    multiply(p1 / (2 * d ** 2), sumMatrices(diagonal00And11)) as Matrix,
    multiply(p1 / (2 * d * Math.sqrt(d)), sumMatrices(coherence00To11)) as Matrix,
    multiply(p2 / (2 * d), sumMatrices(diagonal01And10)) as Matrix,
    multiply(p2 / (2 * d), sumMatrices(coherence01To10)) as Matrix,
  ]);
}

File: python/src/bound_entangled/cd_otimes_cd/badziag_private_singlet.py
"""https://journals.aps.org/prresearch/abstract/10.1103/PhysRevResearch.3.023101 as PPT singlets."""

from toqito.matrices import fourier
from toqito.matrix_ops import tensor
from math import sqrt
from bound_entangled.utils import ketbra
import numpy as np


def ket_i(d: int, i: int):
    ket = np.zeros(d, dtype=np.complex128)
    ket[i] = 1
    return ket


def basis(shield_dim: int, i: int, j: int, k: int, l: int):
    """basis vector |ijkj>"""
    assert i <= 1 and j <= 1
    return tensor(ket_i(2, i), ket_i(2, j), ket_i(shield_dim, k), ket_i(shield_dim, l))


def badziag_private_singlet(*, shield_dim: int) -> np.ndarray:
    """Construct the Bädziąg et al. private-singlet bound-entangled state on C^{2d} ⊗ C^{2d}.

    A PPT singlet in dimension 2d × 2d built from a shield subsystem of
    dimension d and a two-qubit pair, following Phys. Rev. Research 3, 023101
    (2021).  The state is bound entangled for every shield_dim >= 2.

    Args:
        shield_dim: shield subsystem dimension d (>= 2).

    Returns:
        np.ndarray: (4d²) × (4d²) density matrix of the bound-entangled state.

    Examples:
        The matrix is returned in ABA'B' ordering (two-qubit pair A,B then the
        shield pair A',B').  The physical bipartition is Alice=(A,A') against
        Bob=(B,B'), so it has to be permuted to AA'BB' before the Alice|Bob cut
        of dimension 2d x 2d means anything.

        >>> from toqito.matrix_props import is_density
        >>> from toqito.perms import permute_systems
        >>> from toqito.state_props import is_ppt, is_separable
        >>> state = permute_systems(
        ...     badziag_private_singlet(shield_dim=2), [0, 2, 1, 3], dim=[2, 2, 2, 2]
        ... )
        >>> is_density(state)
        True
        >>> is_ppt(state, dim=[4, 4])
        True
        >>> sep, _ = is_separable(state, dim=[4, 4])
        >>> sep
        False
    """
    p1 = sqrt(shield_dim) / (1 + sqrt(shield_dim))
    p2 = 1 - p1
    u = fourier(shield_dim)

    ret = sum(
        ketbra(
            basis(shield_dim, 0, 0, i, j),
        )
        + ketbra(
            basis(shield_dim, 1, 1, i, j),
        )
        for i in range(shield_dim)
        for j in range(shield_dim)
    ) * (p1 / (2 * shield_dim**2))

    ret += sum(
        (
            u[i, j]
            * ketbra(
                basis(shield_dim, 0, 0, i, j),
                basis(shield_dim, 1, 1, j, i),
            )
        )
        + (
            u[i, j].conj()
            * ketbra(
                basis(shield_dim, 1, 1, j, i),
                basis(shield_dim, 0, 0, i, j),
            )
        )
        for i in range(shield_dim)
        for j in range(shield_dim)
    ) * (p1 / (2 * shield_dim * sqrt(shield_dim)))

    ret += sum(
        ketbra(
            basis(shield_dim, 0, 1, i, i),
        )
        + ketbra(
            basis(shield_dim, 1, 0, i, i),
        )
        for i in range(shield_dim)
    ) * (p2 / (2 * shield_dim))

    ret += sum(
        (
            u[i, j]
            * ketbra(
                basis(shield_dim, 0, 1, i, i),
                basis(shield_dim, 1, 0, j, j),
            )
        )
        + (
            u[i, j].conj()
            * ketbra(
                basis(shield_dim, 1, 0, j, j),
                basis(shield_dim, 0, 1, i, i),
            )
        )
        for i in range(shield_dim)
        for j in range(shield_dim)
    ) * (p2 / (2 * shield_dim))

    return ret  # type: ignore
